sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -407,260 +408,284 @@ class GpuSequenceMeta(
+    // will perform an exclusive sum on these values and produce [0, 0, 0, 1, 1].
+    // Combining this with the original predicate boolean array results in the two
+    // T values mapping to indices 0 and 1, respectively.
+    val boolAsInt = withResource(Scalar.fromInt(1)) { one =>
+      withResource(Scalar.fromInt(0)) { zero =>
+        predicate.ifElse(one, zero)
+      }
+    }
+    val prefixSumExclusive = withResource(boolAsInt) { boolsAsInts =>
+      boolsAsInts.scan(ScanAggregation.sum(), ScanType.EXCLUSIVE, NullPolicy.INCLUDE)
+    }
+    val gatherMap = withResource(prefixSumExclusive) { prefixSumExclusive =>
+      // for the entries in the gather map that do not represent valid
+      // values to be gathered, we change the value to -MAX_INT which
+      // will be treated as null values in the gather algorithm
+      withResource(Scalar.fromInt(Int.MinValue)) { outOfBoundsFlag =>
+        predicate.ifElse(prefixSumExclusive, outOfBoundsFlag)
+      }
+    }
+    withResource(gatherMap) { _ =>
+      withResource(new Table(toBeGathered)) { tbl =>
+        withResource(tbl.gather(gatherMap)) { gatherTbl =>
+          gatherTbl.getColumn(0).incRefCount()
+        }
+      }
+    }
+  }
 
   /**
-   * Calculate the size and step (1 or -1) between start and stop both inclusive
-   * size = |stop - start| + 1
-   * step = 1 if stop >= start else -1
-   * @param start first values in the result sequences
-   * @param stop end values in the result sequences
-   * @return (size, step)
+   * Compute the size of each array from 'start', 'stop' and 'step'.
+   * The returned column should be closed.
    */
-  private def calculateSizeAndStep(start: BinaryOperable, stop: BinaryOperable, dt: DataType):
-      Seq[ColumnVector] = {
-    withResource(stop.sub(start)) { difference =>
-      withResource(GpuSequenceUtil.numberScalar(dt, 1)) { one =>
-        val step = withResource(GpuSequenceUtil.numberScalar(dt, -1)) { negativeOne =>
-          withResource(GpuSequenceUtil.numberScalar(dt, 0)) { zero =>
-            withResource(difference.greaterOrEqualTo(zero)) { pred =>
-              pred.ifElse(one, negativeOne)
+  def computeArraySizes(
+      start: ColumnVector,
+      stop: ColumnVector,
+      step: ColumnVector): ColumnVector = {
+    // Keep the same requirement with Spark:
+    //  (step > 0 && start <= stop) || (step < 0 && start >= stop) || (step == 0 && start == stop)
+    withResource(Scalar.fromByte(0.toByte)) { zero =>
+      withResource(new Table(start, stop)) { startStopTable =>
+        // (step > 0 && start <= stop)
+        withResource(step.greaterThan(zero)) { positiveStep =>
+          withResource(startStopTable.filter(positiveStep)) { upTbl =>
+            withResource(upTbl.getColumn(0).lessOrEqualTo(upTbl.getColumn(1))) { allUp =>
+              require(isAllTrue(allUp), "Illegal sequence boundaries: step > 0 but start > stop")
             }
           }
         }
```

Comment:
`positiveStep` should be closed earlier, e.g.:
```suggestion
        val upTbl = withResource(step.greaterThan(zero)) { positiveStep =>
          startStopTable.filter(positiveStep)
        }
        withResource(upTbl) { upTbl =>
          withResource(upTbl.getColumn(0).lessOrEqualTo(upTbl.getColumn(1))) { allUp =>
            require(isAllTrue(allUp), "Illegal sequence boundaries: step > 0 but start > stop")
          }
        }
```

Reply:
Done